hatchet/cdb_reader.py

```
@@ -0,0 +1,180 @@
+        while self.nodes[root_idx].get('parent') != None:
+            root_idx = self.nodes[root_idx].get('parent')
+
+        idx_to_ccnode = {}
+        ccnode_to_metric = {}
+        callpath_to_metrics = {}
+        list_metrics = []
+        list_indices = []
+        pes = set()
+        num_columns = len(self.columns)
+
+        # loop through each data, potentially making node, and then crawling up
+        # each node's parents and potentially making nodes for each parent
+        for data in self.data:
+
+            # get parent of child node indices
+            child = data[node_idx]
+            parent = self.nodes[child].get('parent')
+
+            # count number of processes
+            pes.add(data[self.columns.index('mpi.rank')])
+
+            # skip if this is root
+            if parent == None: continue
+
+            ccchild = idx_to_ccnode.get(child)
+            ccparent = idx_to_ccnode.get(parent)
+
+            if ccparent == None:
+                if ccchild == None:
+                    ccchild = CCNode(_callpath_tuple=None, _parent=None)
+                ccparent = CCNode(_callpath_tuple=None, _parent=None)
+                ccchild.parent = ccparent
+                ccparent.add_child(node=ccchild)
+                ccnode_to_metric[ccparent] = ([self.nodes[parent]['label'], ''] +
+                                              [None] * num_columns)
+            else:
+                if ccchild == None:
+                    ccchild = CCNode(_callpath_tuple=None, _parent=None)
+                ccchild.parent = ccparent
+                if ccchild not in ccparent.children:
+                    ccparent.add_child(node=ccchild)
+
+            # add to mapping for future retrieval
+            idx_to_ccnode[child] = ccchild
+            idx_to_ccnode[parent] = ccparent
+
+            # map metric for child node
+            ccnode_to_metric[ccchild] = [self.nodes[child]['label'], ''] + data
+
+            # walk up parents
```

You shouldn't need to walk up the parents to construct this tree properly -- this makes things _O(avgdepth * n)_.  You can build the tree in just two passes over the data array -- one to build the nodes, and one to hook up each node to its parent.  That should be all you need.  If you find while doing that that any node's parent doesn't exist, you can raise an exception because the file format is invalid.

---

this method and the one that's about to be push to update the PR do not walk up and process nodes that have already been created. it should be an O(n) (n being the number of ccnodes in the calling context tree) algorithm. notice how line 114 it breaks if a parent ccnode has already been created.